ddl/ddl_worker.go

```diff
@@ -149,10 +151,21 @@ func asyncNotify(ch chan struct{}) {
 // buildJobDependence sets the curjob's dependency-ID.
 // The dependency-job's ID must less than the current job's ID, and we need the largest one in the list.
 func buildJobDependence(t *meta.Meta, curJob *model.Job) error {
-	jobs, err := t.GetAllDDLJobs()
+	// Jobs in the same queue are ordered. If we want to find a job's dependency-job, we need to look for
+	// it from the other queue. So if the job is "ActionAddIndex" job, we need find its dependency-job from DefaultJobList.
+	// TODO: rename SetJobListKey to ChangeJobQueue.
+	var jobs []*model.Job
+	var err error
+	switch curJob.Type {
+	case model.ActionAddIndex:
+		jobs, err = t.GetAllDDLJobsInQueue(meta.DefaultJobListKey)
+	default:
+		jobs, err = t.GetAllDDLJobsInQueue(meta.AddIndexJobListKey)
+	}
 	if err != nil {
 		return errors.Trace(err)
 	}
+
 	for _, job := range jobs {
 		if curJob.ID < job.ID {
```

Comment:
Say that there is adding index job of table A, id is 20 and is already enqueued to index job list. And at the same time,  run `alter table A ...`, and the job id is 19, then the job will be enqueued to the general job list.